### DISCUSSION TITLE
Can you access the sqlite database directly ?

### DISCUSSION THREAD

**Discussion by geraldodev** [pocketbase/pocketbase#2590] [Q&A]
Hi, 

Pocketbase is making me to learn go ^^
That's something I didn't expected.

Can I access the database directly  ? (by java for example) I'm assuming that the pocketbase process has exclusive access to the databse and I need to interact via REST.

I'm asking because I don't know go yet, and I wanted to cut some corners, prove an idea and go on...

**Comment by ganigeorgiev** [comment]
Technically if the database is opened in the same WAL mode it should be safe to access the database file directly from multiple applications but keep in mind that only one application/process can write/modify the db file at a time. You may also want to check https://www.sqlite.org/faq.html#q5.

But I personally wouldn't recommend this approach as unexpected issues may arise (eg. requiring you to gracefully handle on your own manually the `SQLITE_BUSY` errors and other locking mechanisms) and instead I'd recommend to use the [web APIs](https://pocketbase.io/docs/api-records/).

Also note that the realtime subscriptions are tied to the PocketBase application process, so when making direct db changes from external app you'll not receive a realtime message for the change (_it will work only if the db change is performed via the web APIs or using the Go `app.Dao()` abstraction_).

**Comment by geraldodev** [comment]
If I make those changes not via webapi but via inprocess go executable that
is using PB as framework, the dangers you describe diminish ?

I want to copy data from another database regularly.

Regards,

Geraldo Lopes de Souza

Em seg., 29 de mai. de 2023 às 03:08, Gani Georgiev <
***@***.***> escreveu:

**Reply by ganigeorgiev** [reply]
> If I make those changes not via webapi but via inprocess go executable that
is using PB as framework, the dangers you describe diminish ?

I'm sorry, I don't understand what do you mean. Could you elaborate what is your planned setup?

If you use PocketBase as framework and have only a single instance running then Yes, the application has builtin fail-retry handling that should minimize the `SQLITE_BUSY` errors. The realtime subscription events will be sent as usual if the record changes are done via the `app.Dao()` abstraction.

If you want to have 2 PocketBase applications operating over a single db file, then you still'll have the auto fail/retry handling but the realtime events will be per application, aka. application A will not receive a realtime event if application B modifies the database.

**Comment by geraldodev** [comment]
The proposed setup would be as follows:
I have one PocketBase (PB) app that utilizes PB as its framework, implemented as a single binary.
The app will retrieve data from another database and generate tables, as well as insert data into the PB database.
Initially, my assumption is that by bypassing the API and writing directly from the PB binary, I could potentially accelerate the table import process.
Furthermore, new tables could be created during this process.

Is it possible to configure the system in a way that only one binary can access the database, thus preventing the situation you described?

**Reply by ganigeorgiev** [reply]
Your described setup should work fine.

> Is it possible to configure the system in a way that only one binary can access the database, thus preventing the situation you described?

I'm not sure, I guess this can be achieved with filesystem lock (eg. making the db file readonly for "other" apps/users), but in your case I don't think it is necessarily.
Even if you have multiple PocketBase instances running on the same server, as long each instance is operating with their own `pb_data` it should be fine (_you can specify an explicit `pb_data` dir with the `--dir` flag; by default the directory is created next to the binary_).
